src/python/pants/engine/README.md

```diff
@@ -76,52 +76,80 @@ In cases where this search detects any ambiguity (generally because there are tw
 that can provide the same product with the same number of parameters), rule graph compilation will
 fail with a useful error message.
 
-### Datatypes
+### How to Declare Engine `Param` Types
 
 In practical use, builtin types like `str` or `int` do not provide enough information to
-disambiguate between various types of data in `@rule` signatures, so declaring small `datatype`
-definitions to provide a unique and descriptive type is highly recommended:
+disambiguate between various types of data in `@rule` signatures, so declaring small, unique classes
+to encapsulate different states can help to make `@rule` sets with complex control flow between rules
+more declarative and self-documenting.
+
+#### Requirements for an Engine `Param`
+
+To use an instance of some class `C` as a `Param` to the engine:
+1. instances of `C` must be immutable,
```

Comment:
Not technically. They _should_ be immutable, but don't need to be. I recommend changing to `should`.

Reply:
If instances of `C` are mutable, then rules can start sharing state by mutating the parameters they're passed in, without notifying the engine that they have done so. We allow that to a very limited extent with the use of `Workspace`, but even that object itself is immutable. Because incorrect behavior may result if `C` is mutable, we would like to state `C` being immutable as a precondition. #8548 tries to ensure this for the common case of `@dataclass` params by ensuring `@frozen_after_init` is used.

I think that `C` needs to be immutable for the engine to work. I'm not sure that we explicitly check that, but #8548 definitely tries to do so. If there is a rewording that might make sense, I'd love to know -- perhaps we could add:
```suggestion
1. instances of `C` must be immutable, (the engine will not be able to verify this)
```

Reply:
True, I agree on why we want to ensure things are immutable.

Maybe:

```
1. instances of `C` must be immutable (although the engine cannot validate this)
```

Why have this disclaimer? It warns plugin authors that the onus is on them to ensure immutability is respected.